### PR TITLE
MCS: set handler parameters when configuring TCBs

### PR DESCRIPTION
**Pull request by corlewis**
This makes it possible to set the badge and rights of fault and timeout handler caps while configuring a TCB.

### CODE REVIEW

File: src/object/tcb.c [hunk 1]
```diff
@@ -1251,17 +1251,31 @@ exception_t decodeSetMCPriority(cap_t cap, word_t length, word_t *buffer)
...
+        thCap = updateCapData(false, thData, thCap);
+        thCap = maskCapRights(rightsFromWord(thRights), thCap);
+    }
+
+    deriveCap_ret_t dc_ret = deriveCap(thSlot, thCap);
+    if (dc_ret.status != EXCEPTION_NONE) {
+        return dc_ret.status;
+    }
+    thCap = dc_ret.cap;
+
```
Comment (Indanz) [inline src/object/tcb.c:1278]: Maybe add a helper function for this block. Except if I missed something, the other two blocks do exactly the same.

**Review by Indanz — COMMENTED** [review]
Overall it seems like an improvement with no downsides.

File: src/object/tcb.c [hunk 2]
```diff
@@ -1251,17 +1251,31 @@ exception_t decodeSetMCPriority(cap_t cap, word_t length, word_t *buffer)
 }
 
 #ifdef CONFIG_KERNEL_MCS
-exception_t decodeSetTimeoutEndpoint(cap_t cap, cte_t *slot)
+exception_t decodeSetTimeoutEndpoint(cap_t cap, word_t length, cte_t *slot, word_t *buffer)
 {
-    if (current_extra_caps.excaprefs[0] == NULL) {
-        userError("TCB SetSchedParams: Truncated message.");
+    if (length < 2 || current_extra_caps.excaprefs[0] == NULL) {
```
[the remark this answers is not in the document]
Comment (corlewis) [reply src/object/tcb.c:1256]: No specific reason from my side, beyond wanting to reuse the existing patterns. It makes verification a lot easier when there is a proof structure that we can reuse.

It's also not immediately obvious to me what it would look like if the parameters are optional. Would it only make sense in this case because they are the only parameters for the function?

File: libsel4/include/interfaces/sel4.xml [hunk 3]
```diff
@@ -372,6 +372,13 @@
                 description="Capability to the scheduling context that the TCB should run on. If the scheduling context is already bound to a notification or TCB that is not this TCB this operation will fail. Similarly, if this TCB is already bound to a scheduling context that is not this scheduling context, this will also fail."/>
             <param dir="in" name="fault_ep" type="seL4_CPtr"
                 description="CPTR to the endpoint which receives IPCs when this thread faults."/>
+            <param dir="in" name="fault_ep_data" type="seL4_Word"
+                description="Optionally set the badge of the new fault handler Endpoint. If both this and fault_ep_rights are set to zero, then this parameter has no effect."/>
```
[the remark this answers is not in the document]
Comment (corlewis) [reply libsel4/include/interfaces/sel4.xml:376]: Thank you for the reminder, I forgot to talk about this design decision. The intention is to both support not modifying the fault handler to maintain compatibility with existing code, and also allow for setting the badge to `0`, which is a valid and I think fairly common value.

Since setting the rights to `0` (i.e. `seL4_NoRights`) would result in an invalid fault handler, I used the combination of both being `0` as a special case to flag not wanting to modify the fault handler cap. With this approach, the badge can still be set to `0` with a combination of `badge=0` and `rights=seL4_AllRights`.

Writing this out has made me realise that an alternative approach could be to just use the rights parameter for this purpose, instead of requiring that both are set to `0`. However, while this would slightly simplify the API, I could see it causing confusion if someone doesn't consider it and tries to set a badge to a specific value without changing the rights parameter away from `0`.

File: src/object/tcb.c [hunk 4]
```diff
@@ -1473,8 +1507,10 @@ exception_t decodeSetSpace(cap_t cap, word_t length, cte_t *slot, word_t *buffer
     }
 
 #ifdef CONFIG_KERNEL_MCS
-    cRootData = getSyscallArg(0, buffer);
-    vRootData = getSyscallArg(1, buffer);
+    word_t fhData    = getSyscallArg(0, buffer);
```
[the remark this answers is not in the document]
Comment (corlewis) [reply src/object/tcb.c:1510]: I have to admit that I had to stare at this for a moment before remembering a reason, but I think it was due to wanting to maintain the same order of parameters as for non-MCS code.

Since the intention is to eventually have both the MCS and non-MCS code support fault handlers in the same way, in some sense I think changing the order of the parameters is unavoidable.

The options appear to me to be
1) change the order for MCS code now
2) change the order for MCS code in the future, when fault handlers are made consistent
3) change the order for non-MCS code in the future, when fault handlers are made consistent

I don't personally have a strong preference for any of these options.

File: src/object/tcb.c (hunk 1 above)
Comment (corlewis) [reply src/object/tcb.c:1278]: Good idea, thank you! That should also make verification easier in the future.

It wouldn't be part of this PR, but for the same reasons I will probably also want to add a helper function for the block that decodes the cspace and vspace parameters. I think that block is the same in `decodeSetSpace` and `decodeTCBConfigure`.

File: libsel4/include/interfaces/sel4.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kent-mcleod) [reply libsel4/include/interfaces/sel4.xml:376]: @corlewis, could you use the fact that every invocation has a length provided by the seL4_MessageInfo_t which you can use to have unambiguously unused parameters by having encodings where the parameters are unused? This is what I believe @Indanz is meaning when talking about optional parameters.

[the remark this answers is not in the document]
Comment (Indanz) [reply libsel4/include/interfaces/sel4.xml:376]: > With this approach, the badge can still be set to 0 with a combination of badge=0 and rights=seL4_AllRights

Only unbadged caps can be rebadged, so the badge value is already 0. In effect the above is a specific corner case of only changing the rights without changing the badge value.

File: src/object/tcb.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Indanz) [reply src/object/tcb.c:1256]: Only reason to not make them optional is to simplify verification, otherwise it always makes sense, if only for backward compatibility. I can imagine that variable length functions are annoying.

To keep backward compatibility in libsel4 the new function needs a new name, like was done with `seL4_SchedControl_Configure` which was renamed to `seL4_SchedControl_ConfigureFlags`.

File: src/object/tcb.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Indanz) [reply src/object/tcb.c:1510]: I'd go for option 3, considering you're changing non-MCS match MCS and not the other way round.

**Comment by Indanz** [comment]
I made a fork of your branch with my suggestions, see [here](https://github.com/seL4/seL4/compare/master...Indanz:seL4:corlewis/set-handler-params).

While working on that I think I found a corner case that is currently not handled correctly: If a badged cap is passed, and another badge value is requested, `updateCapData` will return a null cap, which will result in no fault handler being installed instead of returning an exception.

File: src/object/tcb.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (corlewis) [reply src/object/tcb.c:1256]: I'll explore the verification side of things and if the proofs for the variable length functions are straightforward then I'll be happy to go with your suggested change making these parameters optional.

I don't think it's relevant since it wouldn't help with backward compatibility, but an argument could be made that the existing cspace and vspace parameters are very similar to the two new ones and could be made optional as well. However, I imagine that this would lead to `seL4_TCB_SetSpace` having an overly complicated interface and it probably isn't worth exploring.

File: libsel4/include/interfaces/sel4.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (corlewis) [reply libsel4/include/interfaces/sel4.xml:376]: Right, I explicitly looked at `updateCapData` as part of this work but I then forgot that only unbadged caps could be rebadged. Considering this, I agree with you that the parameters should be independent.

**Comment by corlewis** [comment]
> I made a fork of your branch with my suggestions, see [here](https://github.com/seL4/seL4/compare/master...Indanz:seL4:corlewis/set-handler-params).
> 
> While working on that I think I found a corner case that is currently not handled correctly: If a badged cap is passed, and another badge value is requested, `updateCapData` will return a null cap, which will result in no fault handler being installed instead of returning an exception.
> 
> (Feel free to merge and squash any of these changes into your branch.)

Thank you, I'll first check that they don't make the proofs harder, but I expect that I'll merge most, if not all, of your suggestions.

The corner case you mention is roughly what I was trying to describe in the second paragraph of the Drawbacks section of https://sel4.atlassian.net/browse/RFC-11. I like your suggestion of explicitly returning an exception for this case, but this does feel like it differs from how seL4 handles similar cases in some of the cspace code.

**Comment by corlewis** [comment]
> > While working on that I think I found a corner case that is currently not handled correctly: If a badged cap is passed, and another badge value is requested, `updateCapData` will return a null cap, which will result in no fault handler being installed instead of returning an exception

> The corner case you mention is roughly what I was trying to describe in the second paragraph of the Drawbacks section of https://sel4.atlassian.net/browse/RFC-11. I like your suggestion of explicitly returning an exception for this case, but this does feel like it differs from how seL4 handles similar cases in some of the cspace code.

After looking at the code again, I just realised that all the other examples I was thinking about actually do catch this and return an exception if `updateCapData` or `deriveCap` return a `NullCap`. Based on this, a further change to your suggestion would be to fail if a non-NullCap is passed in and your `update_badge_rights` function results in a `NullCap`.

I think that this would catch all of the corner cases that I am aware of, both the badged cap one you noticed and ones where other types of caps are passed in.

File: src/object/tcb.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Indanz) [reply src/object/tcb.c:1256]: Agreed.

**Comment by Indanz** [comment]
I implemented your suggestion to check for null cap and force pushed to my branch.

**Comment by corlewis** [comment]
> I implemented your suggestion to check for null cap and force pushed to my branch.

Sorry, what I was thinking is a bit more complicated, since a null cap would be valid if it was explicitly passed in. What I wanted to catch is if a non-null cap is passed in but it becomes null after being updated or after calling `deriveCap`.

Initial verification seems ok, so for now I've merged your changes into this PR, with one new commit which adds another helper function and the expanded null cap check.

However, this might be due to my lack of experience with actually building on top of seL4, but I haven't been able to work out a nice way of calling the invocations with optional parameters. @Indanz @kent-mcleod do either of you have a suggestion for the best way to do this?

**Comment by Indanz** [comment]
> I haven't been able to work out a nice way of calling the invocations that makes use of them having optional parameters.

From libsel4's point of view it's just a new system call: Add a new function which also passes the new arguments, and keep the old one as-is?

> Sorry, what I was thinking is a bit more complicated, since a null cap would be valid if it was explicitly passed in. What I wanted to catch is if a non-null cap is passed in but it becomes null after being updated or after calling deriveCap.

Yes, but changing the badge on a null-cap doesn't make any sense, so this distinction seems pointless. With my version you would get an error if you tried that, now you don't. That said, neither version complaints about changing rights on a null-cap. From the kernel's point of view it's not an error as it doesn't care, but for the user it might be a bug.

Moving the `validFaultHandler` check into the helper function is a good idea.

**Comment by corlewis** [comment]
> Yes, but changing the badge on a null-cap doesn't make any sense, so this distinction seems pointless. With my version you would get an error if you tried that, now you don't. That said, neither version complaints about changing rights on a null-cap. From the kernel's point of view it's not an error as it doesn't care, but for the user it might be a bug.

Right, in that case I think we'll need multiple checks to catch all of the corner cases, not just the one that I originally thought we could get away with. What I've been trying to handle is the case where `deriveCap` can return a null cap if for whatever reason certain non-endpoint caps are passed in.

**Comment by Indanz** [comment]
> What I've been trying to handle is the case where deriveCap can return a null cap if for whatever reason certain non-endpoint caps are passed in.

Ah, I didn't think of that one. But that would be two errors: Trying to rebadge an already badged cap, and the cap being of the wrong type. If not rebadging, `validFaultHandler` would catch it as usual. If rebadging, the null-cap check is sufficient.

> in that case I think we'll need multiple checks to catch all of the corner cases, not just the one that I originally thought we could get away with.

I think behaviour should be the same when doing the mutations before calling this function as when doing it within.

**Comment by corlewis** [comment]
> > What I've been trying to handle is the case where deriveCap can return a null cap if for whatever reason certain non-endpoint caps are passed in.
> 
> Ah, I didn't think of that one. But that would be two errors: Trying to rebadge an already badged cap, and the cap being of the wrong type. If not rebadging, `validFaultHandler` would catch it as usual. If rebadging, the null-cap check is sufficient.
> 
> > in that case I think we'll need multiple checks to catch all of the corner cases, not just the one that I originally thought we could get away with.
> 
> I think behaviour should be the same when doing the mutations before calling this function as when doing it within.

I don't think that that's quite right, both because `updateCapData` behaves differently for different types of caps, and because `deriveCap` is called regardless of whether a data/badge value is passed. As an example, if a reply cap is incorrectly passed in then the original null-cap check would not catch it, since `updateCapData` does nothing to reply caps. Following this, `deriveCap` would return a new null cap which finally `validFaultHandler` would accept.

An alternative that I've just thought of is that we could split `validFaultHandler` and first check that the cap passed in is either an endpoint cap or null cap, then mutate it, and then within `updateBadgeRights` check that it is not being badged incorrectly and that it has the correct rights.

**Comment by Indanz** [comment]
Yes, you're right, it's more subtle and harder to get right than it seems. Your alternative sounds reasonable, I would do whatever is easiest for verification. `updateAndCheckHandlerEP` is the only caller of `validFaultHandler`, you could move it closer to each other. They both set the error code now.  _"Mutated cap would be invalid"_ is a very cryptic error the user will only get when trying to rebadge an already badged cap.

**Comment by corlewis** [comment]
I've pushed my first attempt at this [here](https://github.com/corlewis/seL4/commit/2455b4d9235ab2707bdd513a5b9abd8cc2bb173b). I haven't properly tested all of the cases yet, but I think it should handle everything that we have mentioned so far, and should be more clear and consistent with its errors.

**Comment by corlewis** [comment]
> I've pushed my first attempt at this [here](https://github.com/corlewis/seL4/commit/2455b4d9235ab2707bdd513a5b9abd8cc2bb173b). I haven't properly tested all of the cases yet, but I think it should handle everything that we have mentioned so far, and should be more clear and consistent with its errors.

On second thought, the final check could probably just reuse `validFaultHandler`. It would be a bit redundant since we should still know that the cap is either null or an endpoint, but it would probably look a bit nicer and should be slightly easier for verification.

**Comment by Indanz** [comment]
I like it, it's a lot less code overall. I think all corner cases are handled correctly now.

Keeping `validFaultHandler` would be more code, but you could re-use the existing proofs for it.

Nitpicks:
"handler cap not null or an Endpoint"  - > "handler cap not null nor an Endpoint"
"cap already badged or null" -> "handler cap already badged"

**Comment by corlewis** [comment]
> > I haven't been able to work out a nice way of calling the invocations that makes use of them having optional parameters.
> 
> From libsel4's point of view it's just a new system call: Add a new function which also passes the new arguments, and keep the old one as-is?

Following up on this again, either I'm missing something or I don't think that optional parameters like this are a natural fit for seL4 syscalls. My understanding is that the way that the functions are generated from the [xml interface ](https://github.com/seL4/seL4/blob/master/libsel4/include/interfaces/sel4.xml) means that we would need a lot of copied code to support versions of the syscalls with and without the parameters. Even worse, the functions generated from that xml set the `MessageInfo` tag with the unique value of the `method id`, which means that inside seL4 the different syscalls would appear as different invocations.

There is possibly some sort of work around based on representing the optional parameters with a variable length array or variable sized structure, but I'm pretty sure that this would be much more complicated than we would want.

Hopefully someone has a suggestion to get around this, but if not then my preference would be to return to the fixed parameters with `0` values indicating to ignore the parameter. Since @Indanz pointed out that we do not need to support setting the badge to `0` we can still simplify the original implementation in almost all of the ways suggested in this PR. For backwards compatibility we also still have the option of adding a wrapper to `libsel4` which provides the old signature with default values of `0`, similarly to what was done for `seL4_SchedControl_ConfigureFlags`.

**Comment by corlewis** [comment]
This PR has now been updated with compatibility functions, and I think all existing issues and suggestions have been resolved. Verification is also now mostly complete, as seen at https://github.com/seL4/l4v/pull/505. This is currently just for RISCV64 but it is very unlikely that there will be problems on other architectures.

Concrete suggestions for how to adjust the interface to support optional parameters would be very welcome, and would probably allow for the latest commit to be removed. Failing this, naming suggestions for the changed invocations would also be welcomed.

**Comment by kent-mcleod** [comment]
> It's possible to add multiple methods for the same id to libsel4/include/interfaces/sel4.xml with the following change:

@Indanz , do you mean that this should make it possible to generate the calls instead of adding them to [virtual_client.h](https://github.com/seL4/seL4/pull/889/files#diff-27cfe93905ae26f3375f6ac8ecf51e55cb2d1bba941c6265cba9ff97d1f22f10)?

**Comment by Indanz** [comment]
> @Indanz , do you mean that this should make it possible to generate the calls instead of adding them to [virtual_client.h](https://github.com/seL4/seL4/pull/889/files#diff-27cfe93905ae26f3375f6ac8ecf51e55cb2d1bba941c6265cba9ff97d1f22f10)?

Yes.

**Comment by corlewis** [comment]
> It's possible to add multiple methods for the same id to libsel4/include/interfaces/sel4.xml with the following change:

Thank you for the suggestion @Indanz, those changes worked perfectly. The latest commit switches back to the handler parameters being optional.

The only (small) trade-offs remaining of this approach that I am aware of are the largely duplicated methods in `libsel4/include/interfaces/sel4.xml`, and that users of these syscalls now need to consciously consider whether they are setting an endpoint cap or null cap. See https://github.com/corlewis/capdl/commit/cec0350dbc0c271b0123412ea8e3529ee0fc2e84 for an example of the latter, the CapDL loader could previously make use of the default values of 0 for the badges and rights but now needs to explicitly handle the null cap case.

**Comment by Indanz** [comment]
> Thank you for the suggestion @Indanz, those changes worked perfectly.

You're welcome.

> users of these syscalls now need to consciously consider whether they are setting an endpoint cap or null cap.

That's only because the new function doesn't accept a null cap. I think it would work if you change the if check in `updateAndCheckHandlerEP` to:
```
    if (length >= offset + 1 && cap_get_capType(cap) != cap_null_cap) {
```
It seems the intent that the function should accept a null cap.

**Comment by corlewis** [comment]
> That's only because the new function doesn't accept a null cap. I think it would work if you change the if check in `updateAndCheckHandlerEP` to:

To be precise, the problem is due to the combination of passing in a null cap and a badge value. Your suggestion would avoid this, but in an earlier conversation we said that this is a user error that we wanted to catch.

A small alternative is that we could explicitly allow the combination of a null cap and a 0 badge value. That does start to feel a bit overly specific, but would make it possible to keep the CapDL loader example simple while still catching potential errors where a user tries to badge a null cap.

**Comment by Indanz** [comment]
> A small alternative is that we could explicitly allow the combination of a null cap and a 0 badge value. That does start to feel a bit overly specific

But as you said, it makes it easier to write agnostic user space code, and conceptually it makes sense to ignore 0 badge with a null cap. Need to double check that all the subtle corner cases are still handled correctly though, probably a good idea to add tests to sel4test to check them.

**Comment by corlewis** [comment]
> > A small alternative is that we could explicitly allow the combination of a null cap and a 0 badge value. That does start to feel a bit overly specific
> 
> But as you said, it makes it easier to write agnostic user space code, and conceptually it makes sense to ignore 0 badge with a null cap. Need to double check that all the subtle corner cases are still handled correctly though, probably a good idea to add tests to sel4test to check them.

Ok, I've added that check and as hoped for, it did allow nicer user space code in the CapDL loader (https://github.com/seL4/capdl/commit/56633a412727e5e8f77a5be2ed6508e4b82654f9).

As a starting point before adding tests, is this a complete list of the corner cases that we have previously considered?
* Rebadging an already badged endpoint cap
* Rebadging a null cap to a non-0 value
* A non-endpoint, non-null cap is passed in that would be transformed into a null cap by `deriveCap`

**Comment by Indanz** [comment]
Only other one I can think of is:
- Changing rights on a null-cap

File: src/object/tcb.c [hunk 5]
```diff
@@ -1251,22 +1228,89 @@ exception_t decodeSetMCPriority(cap_t cap, word_t length, word_t *buffer)
...
+    default:
+        current_syscall_error.type = seL4_InvalidCapability;
+        return false;
+    }
+    return true;
+}
+
+static deriveCap_ret_t updateAndCheckHandlerEP(cap_t cap, word_t length, cte_t *slot, word_t *buffer,
+                                               word_t offset, word_t capPosition)
+{
```
Comment (axel-h) [inline src/object/tcb.c:1254]: Add sanity check for catching development errors and clarifying function calling contract.
```suggestion
{
    assert( !(buffer && (length > 0) );
    assert( (0 == length) || (offset < length) );
```

File: src/object/tcb.c [hunk 6]
```diff
@@ -1363,12 +1407,12 @@ exception_t decodeSetSchedParams(cap_t cap, word_t length, word_t *buffer)
         return EXCEPTION_SYSCALL_ERROR;
     }
 
-    if (!validFaultHandler(fhCap)) {
-        userError("TCB Configure: fault endpoint cap invalid.");
-        current_syscall_error.type = seL4_InvalidCapability;
-        current_syscall_error.invalidCapNumber = 3;
-        return EXCEPTION_SYSCALL_ERROR;
+    /* fault handler */
+    deriveCap_ret_t dc_ret = updateAndCheckHandlerEP(fhCap, length, fhSlot, buffer, 2, 3);
```
Comment (axel-h) [inline src/object/tcb.c:1411]: Add a `#define DECODE_SET_SHED_PARAMS 2` and use this when checking the length above and pass it here instead of a plain number. Also make the comment state if this is an optional parameter or not.

File: libsel4/include/interfaces/sel4.xml [hunk 7]
```diff
@@ -393,6 +394,54 @@
...
+                    If set to zero, this parameter has no effect.
+                </description>
+            </param>
+            <error name="seL4_IllegalOperation">
+                <description>
+                    The <texttt text="_service"/> is a CPtr to a capability of the wrong type.
+                    Or, <texttt text="_service"/> or <texttt text="sched_context"/> is already bound.
+                    Or, <texttt text="_service"/> is the current thread's TCB.
+                    Or, <texttt text="_service"/> is a TCB in the blocked state and <texttt text="sched_context"/> is not schedulable.
+                </description>
```
Comment (axel-h) [inline libsel4/include/interfaces/sel4.xml:430]: `updateAndCheckHandlerEP()` can also return `seL4_IllegalOperation now also., when `fault_ep_data` cannot be handled ("handler cap being rebadged when already badged or null")
